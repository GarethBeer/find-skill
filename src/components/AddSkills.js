import React, { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCheckCircle,
  faExclamationCircle,
} from "@fortawesome/free-solid-svg-icons";
import Alert from "./Alert";
import { postSkill } from "../requests/requests";
import "../styles/AddSkills.css";

const Skills = () => {
  const initialState = {
    fields: {
      name: "",
      skill: "",
      description: "",
      postcode: "",
      free: true,
      professional: false,
      email: "",
    },
    alert: {
      message: "",
      success: false,
    },
    validationErrors: {},
  };
  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);
  const [validationErrors, setvalidationErrors] = useState(
    initialState.validationErrors
  );
  // POST skill
  const handleValidation = () => {
    const errors = {};
    let formIsValid = true;

    if (!fields.name) {
      formIsValid = false;
      errors.name = "Required field";
    }
    if (!fields.postcode) {
      formIsValid = false;
      errors.postcode = "Required field";
    }

    if (!fields.postcode.match(/[A-Z]{1,2}[0-9]{1,2}[A-Z]?\s?[0-9][A-Z]{2}/i)) {
      formIsValid = false;
      errors.postcode = "Invalid postcode";
    }

    if (!fields.skill) {
      formIsValid = false;
      errors.skill = "Required field";
    }
    if (!fields.description) {
      formIsValid = false;
      errors.description = "Required field";
    }
    if (!fields.email) {
      formIsValid = false;
      errors.email = "Required field";
    }
    setvalidationErrors(errors);
    return formIsValid;
  };

  const handleAddSkill = (event) => {
    event.preventDefault();
    setAlert({ message: "", isSuccess: false });
    if (handleValidation()) {
      postSkill(fields, setAlert);
    }
  };

  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  return (
    <div>
      <Alert message={alert.message} success={alert.isSuccess} />
      <div className="AddSkill">
        <h3>Add a skill</h3>
        <form className="add-form" onSubmit={handleAddSkill}>
          <label className="label" htmlFor="name">
            Your name:
            <input
              className="input"
              type="text"
              placeholder="Add your name"
              id="name"
              name="name"
              value={fields.name}
              onChange={handleFieldChange}
            />
            {!validationErrors.name && fields.name && (
              <FontAwesomeIcon className="okIcon" icon={faCheckCircle} />
            )}
            {validationErrors.name && (
              <FontAwesomeIcon
                className="notOkIcon"
                icon={faExclamationCircle}
              />
            )}
          </label>

          <label className="label" htmlFor="skill">
            Skill title:
            <input
              className="input"
              type="text"
              placeholder="Short desrcription of your skill e.g. 'Fixing Bikes'"
              id="skill"
              name="skill"
              value={fields.skill}
              onChange={handleFieldChange}
            />
            {!validationErrors.skill && fields.skill && (
              <FontAwesomeIcon className="okIcon" icon={faCheckCircle} />
            )}
            {validationErrors.skill && (
              <FontAwesomeIcon
                className="notOkIcon"
                icon={faExclamationCircle}
              />
            )}
          </label>

          <label className="label" htmlFor="description">
            Description:
            <input
              className="input"
              type="text"
              placeholder="Describe your skills with few sentences"
              id="description"
              name="description"
              value={fields.description}
              onChange={handleFieldChange}
            />
            {!validationErrors.description && fields.description && (
              <FontAwesomeIcon className="okIcon" icon={faCheckCircle} />
            )}
            {validationErrors.description && (
              <FontAwesomeIcon
                className="notOkIcon"
                icon={faExclamationCircle}
              />
            )}
          </label>

          <label className="label" htmlFor="postcode">
            Postcode:
            <input
              className="input"
              type="text"
              placeholder="Add your postcode e.g 'SK17 7DW'"
              id="postcode"
              name="postcode"
              value={fields.postcode}
              onChange={handleFieldChange}
            />
            {!validationErrors.postcode && fields.postcode && (
              <FontAwesomeIcon className="okIcon" icon={faCheckCircle} />
            )}
            {validationErrors.postcode && (
              <FontAwesomeIcon
                className="notOkIcon"
                icon={faExclamationCircle}
              />
            )}
          </label>

          <label className="label" htmlFor="free">
            Free Service:
            <select
              title="free"
              className="select"
              id="free"
              name="free"
              value={fields.free}
              onChange={handleFieldChange}
            >
              <option value="true">Yes</option>
              <option value="false">For fee</option>
            </select>
          </label>

          <label className="label" htmlFor="professional">
            Professional service:
            <select
              title="professional"
              className="select"
              id="professional"
              name="professional"
              value={fields.professional}
              onChange={handleFieldChange}
            >
              <option value="true">Yes</option>
              <option value="false">No</option>
            </select>
          </label>

          <label className="label" htmlFor="email">
            Contact email address:
            <input
              className="input"
              type="email"
              placeholder="Add your email address"
              id="email"
              name="email"
              value={fields.email}
              onChange={handleFieldChange}
            />
            {!validationErrors.email && fields.email && (
              <FontAwesomeIcon className="okIcon" icon={faCheckCircle} />
            )}
            {validationErrors.email && (
              <FontAwesomeIcon
                className="notOkIcon"
                icon={faExclamationCircle}
              />
            )}
          </label>

          <button className="add-button" type="submit">
            Add
          </button>
        </form>
      </div>
    </div>
  );
};

export default Skills;
