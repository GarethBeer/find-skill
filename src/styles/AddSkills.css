.manage-skill {
  display: flex;
  width: 100%;
}
.add-skill {
  width: 48%;
}
.add-form {
  display: flex;
  flex-direction: column;
  overflow-y: scroll;
  padding-bottom: 1.25rem;
}
h3 {
  margin-left: 2vh;
}
.input, .select {
    color: rgb(10, 31, 10) ;
    width: 90%;
    border: 1px solid  rgb(5, 107, 5);
    margin: 0.5% 1% 1.5% 1%;
    padding: 1.5%;
    box-sizing: border-box;
    font-weight: 600;
    background-color:ivory;
    border-radius: 4px;
}
.okIcon {
    color: lawngreen;
}
.notOkIcon {
    color: rgb(121, 4, 4);
}

.add-button {
    background-color:ivory;
    border-radius: 4px;
    border: 1px solid rgb(9, 32, 2);
    cursor: pointer;
    padding: 1.5% 10%;
    font-weight: 600;
    margin-left: 3%;
    width: 30%;
  }

.label{
    padding-left: 1vw;
}

.add-button:hover {
    background-color: lawngreen;
  }

::placeholder {
    color: rgb(10, 31, 10);
}
.change-skill {
  background-color: rgb(90, 153, 27);
  width: 48%;
}
.img{
  width: 100%;
}
h2 {
  padding: 1vw ;
  text-align: justify;
  margin: 0em;
}
@media only screen and (max-width: 732px) {
  .manage-skill {
    flex-direction: column;
  }
  .add-skill {
    width: 100%;
    order: 2;
  }
  .change-skill {
    width: 100%;
    order: 1;
  }
  .input, .select {
    font-weight: 300;
    font-size: 12px;
  }
  h2 {
    font-size: 14px;
  }
}