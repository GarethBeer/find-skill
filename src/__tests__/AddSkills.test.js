import React from "react";
import { render, fireEvent } from "@testing-library/react";
import AddSkills from "../components/AddSkills";

describe("AddSkill", () => {
  const setUp = () => {
    const { asFragment, getByRole } = render(<AddSkills />);

    return {
      asFragment: asFragment(),
      name: getByRole("textbox", { name: /name/i }),
      skill: getByRole("textbox", { name: /skill/i }),
      description: getByRole("textbox", { name: /description/i }),
      postcode: getByRole("textbox", { name: /postcode/i }),
      email: getByRole("textbox", { name: /email/i }),
      free: getByRole("combobox", { name: /free/i }),
      professional: getByRole("combobox", { name: /professional/i }),
      addButton: getByRole("button", { name: /add/i }),
    };
  };

  const mockHandleFieldChange = (field, value) => {
    fireEvent.change(field, {
      target: { value },
    });
  };

  let renderedComponent;

  beforeEach(() => {
    renderedComponent = setUp();
  });

  describe("renders correctly", () => {
    it("matches the AddSkills snapshot", () => {
      expect(renderedComponent.asFragment).toMatchSnapshot();
    });

    it("name", () => {
      expect(renderedComponent.name).toHaveClass("input");
      expect(renderedComponent.name).toHaveAttribute("type", "text");
      expect(renderedComponent.name).toHaveAttribute("id", "name");
      expect(renderedComponent.name).toHaveAttribute("name", "name");
    });

    it("skill", () => {
      expect(renderedComponent.skill).toHaveClass("input");
      expect(renderedComponent.skill).toHaveAttribute("type", "text");
      expect(renderedComponent.skill).toHaveAttribute("id", "skill");
      expect(renderedComponent.skill).toHaveAttribute("name", "skill");
    });

    it("description", () => {
      expect(renderedComponent.description).toHaveClass("input");
      expect(renderedComponent.description).toHaveAttribute("type", "text");
      expect(renderedComponent.description).toHaveAttribute(
        "id",
        "description"
      );
      expect(renderedComponent.description).toHaveAttribute(
        "name",
        "description"
      );
    });

    it("postcode", () => {
      expect(renderedComponent.postcode).toHaveClass("input");
      expect(renderedComponent.postcode).toHaveAttribute("type", "text");
      expect(renderedComponent.postcode).toHaveAttribute("id", "postcode");
      expect(renderedComponent.postcode).toHaveAttribute("name", "postcode");
    });

    it("email", () => {
      expect(renderedComponent.email).toHaveClass("input");
      expect(renderedComponent.email).toHaveAttribute("type", "text");
      expect(renderedComponent.email).toHaveAttribute("id", "email");
      expect(renderedComponent.email).toHaveAttribute("name", "email");
    });

    it("free", () => {
      expect(renderedComponent.free).toHaveClass("select");
      expect(renderedComponent.free).toHaveAttribute("id", "free");
      expect(renderedComponent.free).toHaveAttribute("name", "free");
    });

    it("professional", () => {
      expect(renderedComponent.professional).toHaveClass("select");
      expect(renderedComponent.professional).toHaveAttribute(
        "id",
        "professional"
      );
      expect(renderedComponent.professional).toHaveAttribute(
        "name",
        "professional"
      );
    });
  });

  describe("form fields works correctly onChange", () => {
    it("name", () => {
      expect(renderedComponent.name.value).toBe("");
      mockHandleFieldChange(renderedComponent.name, "Test_Name");
      expect(renderedComponent.name.value).toBe("Test_Name");
    });

    it("skill", () => {
      expect(renderedComponent.skill.value).toBe("");
      mockHandleFieldChange(renderedComponent.skill, "Test_Skill");
      expect(renderedComponent.skill.value).toBe("Test_Skill");
    });

    it("description", () => {
      expect(renderedComponent.description.value).toBe("");
      mockHandleFieldChange(renderedComponent.description, "Test_describe");
      expect(renderedComponent.description.value).toBe("Test_describe");
    });

    it("postcode", () => {
      expect(renderedComponent.postcode.value).toBe("");
      mockHandleFieldChange(renderedComponent.postcode, "Test_postcode");
      expect(renderedComponent.postcode.value).toBe("Test_postcode");
    });
    it("free", () => {
      expect(renderedComponent.free.value).toBe("true");
      mockHandleFieldChange(renderedComponent.free, "false");
      expect(renderedComponent.free.value).toBe("false");
    });

    it("professional", () => {
      expect(renderedComponent.professional.value).toBe("false");
      mockHandleFieldChange(renderedComponent.professional, "true");
      expect(renderedComponent.professional.value).toBe("true");
    });

    it("email", () => {
      expect(renderedComponent.email.value).toBe("");
      mockHandleFieldChange(renderedComponent.email, "Test_email@com");
      expect(renderedComponent.email.value).toBe("Test_email@com");
    });
  });
});
