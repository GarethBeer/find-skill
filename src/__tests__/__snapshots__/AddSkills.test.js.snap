// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSkill renders correctly matches the AddSkills snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="AddSkill"
    >
      <h3>
        Add a skill
      </h3>
      <form
        class="add-form"
      >
        <label
          class="label"
          for="name"
        >
          Your name:
          <input
            class="input"
            id="name"
            name="name"
            placeholder="Add your name"
            type="text"
            value=""
          />
        </label>
        <label
          class="label"
          for="skill"
        >
          Skill title:
          <input
            class="input"
            id="skill"
            name="skill"
            placeholder="Short desrcription of your skill e.g. 'Fixing Bikes'"
            type="text"
            value=""
          />
        </label>
        <label
          class="label"
          for="description"
        >
          Description:
          <input
            class="input"
            id="description"
            name="description"
            placeholder="Describe your skills with few sentences"
            type="text"
            value=""
          />
        </label>
        <label
          class="label"
          for="postcode"
        >
          Postcode:
          <input
            class="input"
            id="postcode"
            name="postcode"
            placeholder="Add your postcode e.g 'SK17 7DW'"
            type="text"
            value=""
          />
        </label>
        <label
          class="label"
          for="free"
        >
          Free Service:
          <select
            class="select"
            id="free"
            name="free"
            title="free"
          >
            <option
              value="true"
            >
              Yes
            </option>
            <option
              value="false"
            >
              For fee
            </option>
          </select>
        </label>
        <label
          class="label"
          for="professional"
        >
          Professional service:
          <select
            class="select"
            id="professional"
            name="professional"
            title="professional"
          >
            <option
              value="true"
            >
              Yes
            </option>
            <option
              value="false"
            >
              No
            </option>
          </select>
        </label>
        <label
          class="label"
          for="email"
        >
          Contact email address:
          <input
            class="input"
            id="email"
            name="email"
            placeholder="Add your email address"
            type="email"
            value=""
          />
        </label>
        <button
          class="add-button"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
